### ISSUE TITLE
Filters and pandoc-citeproc in docx template

### ISSUE THREAD

**Issue by talazem** [htdebeer/pandocomatic#36]
I have a Pandocomatic issue whilst using the `authorRemoveHash` filter when outputting to docx, using citeproc. 

I may be misunderstanding this, but — in the `docx-with-refs` template — I must repeat my invoking of the pandoc-crossref, pandoc-citeproc, and pandoc-citeproc-preamble filters, even though they were invoked in the `refs` template above. Without doing this, if I invoked the `removeHR` and `authorRemoveHash` filters, citeproc was not working.

The following are the relevant parts of the pandocomatic.yaml file. In `docx-with-refs` I have escaped the three filters already invoked by `refs`. This results in the error described. Unescape them, and everything works as expected.

settings:
recursive: true
follow-symlinks: false
skip: ['.*', 'pandocomatic.yaml']
templates:
#######################################
refs:
pandoc:
standalone: true
filter: 
\- pandoc-crossref
\- pandoc-citeproc #process citations
\- pandoc-citeproc-preamble
bibliography: /Users/name/.pandoc/Bibliography.bib
csl: /Users/name/.pandoc/csl/chicago-note-bibliography.csl
number-sections: true # required by pandoc-crossref for LaTeX/PDF output, recommended for other output formats
metadata:
notes-after-punctuation: true
link-citations: true 
#######################################
docx:
setup: []
preprocessors: []
pandoc:
from: markdown
to: docx
standalone: true
dpi: 300
reference-docx: '/Users/name/.pandoc/templates/reference.docx'
number-sections: true # required by pandoc-crossref for LaTeX/PDF output, recommended for others
postprocessors: []
cleanup: []
#######################################
docx-with-refs:
extends: ['refs','docx']
pandoc:
filter: 
\# - pandoc-crossref
\# - pandoc-citeproc #process citations
\# - pandoc-citeproc-preamble
\- '/Users/name/.pandoc/filters/removeHR' #remove horizontal rules
\- '/Users/name/.pandoc/filters/authorRemoveHash' #converts author metadata for Word
#######################################

There may be other dependencies I am overlooking, but perhaps this is a starting point.

**Comment by iandol** [comment]
Just to clarify, these filters (`authorRemoveHash` and `removeHR`) are not part of pandocomatic, but user filters. The issue seems to be, with TEMPLATEA being the parent of TEMPLATEC that the inheritance of the filters from A fails if C specifies extra filters? I will try and test this (though I can't do this before a few days) and see if I can reproduce.

**Comment by htdebeer** [comment]
I'll have a look at this in the weekend.

It might have something to do with merging "list parameters" (or the lack thereof). Is the same thing happening for, say, the `--css` parameters: if you'd set CSS files in the parent and in the child, does the CSS files of the parent show up when using the child template?

Fundamentally, how should list parameters be merged when extending a template? Should they? Should you be able to remove or overwrite list elements from the parent?

**Comment by iandol** [comment]
I made a simple test case and can confirm this behaviour:

```yaml
#######################################     
  test-A:    
    pandoc:
      from: markdown
      to: html5
      standalone: true
      css: '/Users/ian/.pandoc/templates/custom.css'
      filter: '/Users/ian/.pandoc/filters/noop'
 #######################################     
  test-B:    
    pandoc:
      filter: '/Users/ian/.pandoc/filters/noopB'
 #######################################     
  test-C:   
    extends: ['test-A','test-B'] 
    pandoc:
      css: '/Users/ian/.pandoc/templates/customC.css'
      filter: '/Users/ian/.pandoc/filters/noopC'
```

```markdown
---
title: "Test"
pandocomatic:
  use-template: test-C
---

# Abstract #

Blah blah blah.
```

```
👉  pandocomatic -b -i 'Lu.md'
pandoc	--from=markdown \
	--to=html5 \
	--standalone \
	--css=/Users/ian/.pandoc/templates/customC.css \
	--filter=/Users/ian/.pandoc/filters/noopC
Pandocomatic needed 0.8 seconds to convert 'Lu.md'.
```
So yes, the parent filters and CSS are lost and overwritten by the child items. I had naively assumed for list parameters they were merged, but in my templates never needed to use merging so never saw this when I inherited from parents (my parent only ever contains pandoc-citeproc, which is called even if you don't call it when you have --Bibliography defined). I think merging is generally more flexible than replacing, but the problem is if you do want to override the parent, how can you do that? In that case the benefit of replacement is you can handle merge (manually) and replace. This is less elegant but more practical. 

```yaml
#######################################     
  test-C:   
    extends: ['test-A','test-B'] 
    pandoc:
      css: 
        - 'replaceListItems!'
        - '/Users/ian/.pandoc/templates/customC.css' # this replaces
      filter: '/Users/ian/.pandoc/filters/noopC' # this merges
```

One possible solution is to append a custom tag, and pandocomatic would parse the yaml to determine items to merge/replace, but though the best of both worlds it is inelegant IMO.

**Comment by htdebeer** [comment]
As another solution, for these "list parameters" I can introduce options like:

- add_css, add_filter, ...
- remove_css, remove_filter, ...

in pandocomatic.

The question then is, what should the options/methods without a prefix do? As it is now: replacing the whole property?

It would be different from pandoc and paru, though, so that might be confusing for some users.

**Comment by talazem** [comment]
My apologies for what might be a naive contribution, as I know nothing about the logic of coding such things. But to mind, what I understood was that a template X that contains ‘extends:’ would inherit all the parameters of the extended templates (say, A and B). Anything parameters specific to template X would either be additions that were not specified before, and by definition would not be in conflict; or, if they were previously specified, would be understood to override them. So if template X was inheriting ‘to: latex’ from template B, but then specifies ‘to: PDF’, the latter would override the former.

I am sure this is obvious, but I just thought to demonstrate what I, and perhaps others, expect to happen when we speak of ‘extending’ or inheriting features.

**Comment by htdebeer** [comment]
Don't sell yourself short. It is through the constructive contribution from users that free software can improve beyond a mere personal side-project.

Anyway, if we look at pandocomatic templates, there are three types of properties to take into account when extending a template:

1. simple values, like numbers, strings, and booleans,
2. maps of any value, and
3. lists of any value.

Given templates `A` and `B`, `B` extends `A` on property x, then the following cases seem relevant:

# 1. Simple values

For simple values, extension means overwriting: if `B` =>`x: 4` extends `A`=>`x: true`, then the result should be `x: 4`.

But how can we remove property `A.x` in `B`?

# 2. Map values

For map values, two cases can occur:

1.  Extension means adding extra properties to the map: if `B` => `x: {b: 2}` extends `A`=>`x: {a: 1}`, then the result should be `x: {a: 1, b: 2}`.
2. Extension means overwriting properties in the map: if `B' => `x: {a: 2}` extends `A` => `x: {a:1, b: 2}`, then the result should be `x: {a: 2, b : 2}`.

But how can we remove property `A.x.a` in `B`? And what if we want to remove or replace property `A.x` all together in `B`?

# 3. List values

For list values, the situation is a bit different because we do not know if the list is a [set](https://en.wikipedia.org/wiki/Set_(mathematics)) or a [multiset](https://en.wikipedia.org/wiki/Multiset). That is: are values allowed to occur more than once in a list property?

1. Extension means adding values to the list: if `B` => `x: [1, 2]` extends `A` => `x: [3]`, then the result should be `x: [1, 2, 3]`. However,
2. Extension means adding values to the list: if `B` => `x: [1, 2]` extends `A` => `x: [1]`, then the result should be `x: [1, 2, 1]` (treated as a multiset)? Or should it be `x: [1, 2]` (treated as a set)?

Now I am thinking about it, it seems that most pandocomatic and pandoc list properties act as sets. That would mean that extension means adding or replacing values to keep the values in a list unique.

But how can we remove a value from `A.x` in `B`?  And what if we want to remove or replace property `A.x` all together in `B`?

# Removing values

The extension mechanism seems to favor adding or replacing values. Removing values or replacing map or list properties on the other hand does not seem to be covered. If we want to support removal and replacing during extension, we need to come up with some mechanism to define these actions in the YAML metadata.

For removal of properties, using a NULL value of sorts could do the trick. If template `B` sets property `x` to value `NULL`, that property is removed from the result.

This does not help by removing values from a list, however. In the examples above. How can we specify we do not want value `1` in the resulting list `x`? Two possible solutions are:

1. Replace a whole list with the unwanted values removed. But to differentiate between "additive" extension, a modifier of sorts is needed.
2. Remove a value by means of a modifier.

In both solutions, a modifier is needed. Ian suggested the exclamation mark (`!`). But that would mean that each value has to be a string, for otherwise the exclamation mark interferes with the parsing of the value. Another option is to use a map instead of a list when using an "subtractive" extension. For example, given `A` => `x: [1, 3]`:

`B` => `x: {replace: true, value: [2,3]}`, to result in `x: [2, 3]`

or

`B` => `x: {remove: [1], add: [5]}`, to result in `x: [2, 3, 5]`.

**Comment by iandol** [comment]
Huub, a gracious, flexible and receptive developer like you also makes a huge difference! 😸  🥇 

## Removing items...
YAML does specify that empty nodes are _commonly_ treated as "null":

http://www.yaml.org/spec/1.2/spec.html#id2786563 

So, in this case removal would most simply be specified for values and maps:

* Simple: `B` => `x: ` extends `A` => `x: 1` — should remove `x` from `A`
* Map:  `B` => `x: {a:  }` extends `A` => `x: {a: 1}` — should remove `a` from `x` from `A`
* Map:  `B` => `x: ` extends `A` => `x: {a: 1}` — should remove `x` from `A`

### Dealing with lists...
For lists, I think you are right that they are most well modelled as **sets** for Pandoc; e.g. for filters, CSS and other lists of multiple items it would be detrimental to add the same item twice. Metadata like author or keyword lists also do not need a multi-set representation. So `B` => `x: [1, 2]` extends `A` => `x: [1]` should yield `B` => `x: [1, 2]` IMO!

Yes, the ! could only be used for key names, and you are right that therefore it could not be used for list items; but it can still be used for the list key. You idea to use a map to pass values to manipulate the list is _very neat_! If this was easy to accomplish it would be the nicest way. But otherwise I think the simple route is to not allow items in a list to be modified and only the whole list to be replaced. To replace the whole list we still need to use something like ! on the key (I think keys are always going to be strings):

* `B` => `x: [2, 3]` extends `A` => `x: [1]` — should yield `x: [2, 3, 1]`
* `B` => `x!: [2, 3]` extends `A` => `x: [1]` — should yield `x: [2, 3]`

In summary I think your map idea `{replace: , value: , add: , remove: }` would be the most elegant, but a simpler replace method (if it would be simpler to implement) would also be workable. 

-----------------------------

Ad an aside; YAML does allow you to duplicate data (&tag and then *tag) and selectively replace items; I looked to see if this could be useful, though I don't think it can help us in any way

https://en.wikipedia.org/wiki/YAML#Advanced_components

**Comment by htdebeer** [comment]
I didn't think to look into the YAML spec. Thanks for taking that step, for using an "empty" property to remove it is very neat!

As for the difficulty of implementing using an update object for lists, that shouldn't be too much of a problem. Extension flows from parent to child, so if the parent property is a list, and the child has the same property, it is easy enough to match property types and choose a different behavior for list<->list and list<->map.

I'll free up some development time for this next week. It is a nice addition to version 0.2 of pandocomatic.

**Comment by htdebeer** [comment]
I resolved this issue in version 0.1.4.16: you can add/remove items in a list by using a map with the properties 'add' and 'remove'.

**Comment by talazem** [comment]
Thank you for this. I've updated to 0.1.4.17.

I'd be grateful for a brief example, perhaps using the templates which I sent you when originally filing this problem.

Since updating, I am getting an error when using Pandoc(omatic) to convert a pandoc/MD document to LaTeX:

Error running pandoc => error while running:

```
pandoc	--to=latex \
	--template=/Users/Talal/.pandoc/templates/custom-memoir.latex \
	--table-of-contents \
	--base-header-level=1 \
	--standalone \
	--filter=/usr/local/bin/pandoc-crossref \
	--filter=/usr/local/bin/pandoc-citeproc \
	--filter=/Users/Talal/.cabal/bin/pandoc-citeproc-preamble \
	--filter=/Users/Talal/.pandoc/filters/removeHR \
	--filter=/Users/Talal/.pandoc/filters/authorToName \
	--filter=/Users/Talal/.pandoc/filters/pandoc-word-pagebreak \
	--bibliography=/Users/Talal/.pandoc/Bibliography.bib \
	--csl=/Users/Talal/.pandoc/csl/chicago-note-bibliography.csl \
	--number-sections \
	--from=markdown \
	--latex-engine=xelatex \
	--wrap=preserve \
	--dpi=300 \
	--output=./presentation.pdf

Pandoc responded with:

! LaTeX Error: Missing \begin{document}.

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.81 \textbackslash
```

When I open up my LaTeX file, I find the following offending lines:

> \textbackslash{}AtBeginDocument\{\%
> 
> 
> \}
> \textbackslash{}AtBeginDocument\{\%
> 
> 
> \}
> 
> \textbackslash{}AtBeginDocument\{\%
> 
> \}

I'd be grateful for your clarification.

**Comment by talazem** [comment]
I should have added the line that came before the first offending line:

```
\usepackage{subfig}
\textbackslash{}AtBeginDocument\{\%
```

Ian, let me know if this insertion of the subfig package is due to one of the filters we are using. Otherwise, I'm not sure why this has happened only after upgrading pandocomatic (I was getting no errors just 10 minutes before).

**Comment by talazem** [comment]
(With apologies for the multiple emails):

I've now determined that the error is arising when using Ian's ‘authortoname’ filter. When I switch it off, I don't get the error.

This filter was not causing this error before upgrading Pandocomatic. So I imagine it is simply a matter my learning of how to use the properties ‘add’ and ‘remove’ on filters:

```
      filter: 
        - '/Users/Talal/.pandoc/filters/authorToName' #convert author to author:
```

**Comment by htdebeer** [comment]
You can find an example [here](https://github.com/htdebeer/pandocomatic/blob/master/example/twice_extended_wiki.yaml).

If the problem persists, please submit a complete yet minimal example of (test) files and a pandocomatic invocation to reproduce the issue. I then can try to reproduce it here and fix it.

**Comment by talazem** [comment]
Excellent. Using the nested ‘add:’, it all works perfectly. Sorry for the false alarm.

**Comment by iandol** [comment]
Glad it got fixed! I'm surprised the filters would cause the LaTeX changes you found, they shouldn't do that, so I hope it was just some glitch. LaTeX errors can often be incredibly difficult to understand, it is such an overly complex stack of tools (and decades of patches on top of patches). I think the newer HTML>PDF tools are much more logical, if you can achieve your layout in CSS then you can generate good PDF output (perhaps with Maths as the last domain where LaTeX will always dominate)...
